backends/arm/runtime/ArmBackendEthosU.cpp

```diff
@@ -108,18 +103,93 @@ class ArmBackend final : public PyTorchBackendInterface {
         handles.scratch_data,
         handles.scratch_data_size);
 
-    // Write inputs into SRAM scratch area defined by Vela
+    // Write argument values (from EValue tensor) into Ethos-U scratch
+    // TODO(MLETORCH-123): Optimise into direct write from Vela into the SRAM
+    //                     or DRAM output for compatible data layouts.
     for (int i = 0; i < handles.inputs->count; i++) {
-      const char* input_addr =
-          handles.scratch_data + handles.inputs->io[i].offset;
-      // Process input EValue into scratch
-      // TODO: Optimise into direct write from Vela into the SRAM or DRAM output
-      //       for compatible data layouts.
-      int* input_address = (int*)input_addr;
       auto tensor_in = args[i]->toTensor();
-      for (int j = 0; j < tensor_in.numel(); j++) {
-        // TODO: extend beyond tensors with 4 byte elements
-        input_address[j] = tensor_in.mutable_data_ptr<int>()[j];
+      VelaIO* scratch_in = &handles.inputs->io[i];
+      char* scratch_addr = handles.scratch_data + handles.inputs->io[i].offset;
+
+      // We accept:
+      bool supported = 0;
+      // 32 bit int (simple non-quantised test cases)
+      supported |=
+          (tensor_in.scalar_type() == ScalarType::Int and
+           handles.inputs->io[i].elem_size == 4);
+      // 8 bit int (IOQDQ pass prepared networks)
+      supported |=
+          (tensor_in.scalar_type() == ScalarType::Char and
+           handles.inputs->io[i].elem_size == 1);
+      if (!supported) {
+        ET_LOG(
+            Error,
+            "Input %d expected Integer (4 byte) or Char (1 byte) integer inputs",
+            i);
+        return Error::InvalidProgram;
+      }
+
+      // Special case implicitly permuted rank4 tensors
+      int permuted_input_shape = false;
```

Comment:
nit: can we refactor the permute stuff out as a function?

Reply:
Sure can! ;)

Reply:
this was refactored out.